qualtran/bloqs/comparison_gates.ipynb

```diff
@@ -0,0 +1,957 @@
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As we did before we construct the gate to compare two 2-qubit numbers. First we print the decomposition of the gate followed by the result of running the gate on each of 16 possiblities individually and finally the result of running the gate on the uniform superposition of all 16 possibilities."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "first_quantum_number = cirq.NamedQubit.range(\n",
+    "    2, prefix='P'\n",
+    ")  # The qubit that will hold quantum numbers.\n",
+    "second_quantum_number = cirq.NamedQubit.range(\n",
+    "    2, prefix='Q'\n",
+    ")  # The qubit that will hold quantum numbers.\n",
+    "quantum_quantum_equality = cirq.Circuit(\n",
+    "    equality_oracle_quantum_quantum(first_quantum_number, second_quantum_number, z)\n",
+    ")\n",
+    "quantum_quantum_equality"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Now we check individual possiblities.\n",
+    "check_each_possibility(\n",
+    "    quantum_quantum_equality,\n",
+    "    first_quantum_number + second_quantum_number,\n",
+    "    [z],\n",
+    "    quantum_classical=False,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Finally we check the uniform super position.\n",
+    "check_uniform_superposition(\n",
+    "    quantum_quantum_equality,\n",
+    "    first_quantum_number + second_quantum_number,\n",
+    "    [z],\n",
+    "    quantum_classical=False,\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Notice that the ancilla qubit are always clean at the end of execution and that the input qubits are not affected**"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "q8Zyy7SVtnxo"
+   },
+   "source": [
+    "## The quantum-classical comparator with $4n + \\mathcal{O}(1)$ T gates\n",
+    "\n",
+    "### Inspiration\n",
+    "We will only consider the comparison oracles for less than since the $\\leq$ oracle has one extra clifford operation (a CNOT) and the greater than oracle has exactly 2 extra clifford operations, the same CNOT and an X.\n",
+    "\n",
+    "While [Berry et al., 2019](https://doi.org/10.48550/arXiv.1902.02134) reduced the problem to subtraction, we take a different path and noting that comparing two numbers of equal size is the same finding which of them is lexicographically smaller, this problem is usually solved sequentially and is essentially a finite state machine of $n + 3$ states each having two transitions. The result is an almost identical decomposition (up to clifford operations) with the same T complexity.\n",
```

Comment:
big run on sentence with at least one grammar problem! Please refactor into more than one sentence.

suggest:
 - we take a different path.
 - We note that .. is the same **as** ... smaller.
 - This problem is usually solved ...